src/ucs/type/rwlock.h

```diff
@@ -0,0 +1,108 @@
+/*
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_RWLOCK_H
+#define UCS_RWLOCK_H
+
+#include <sched.h>
+#include <errno.h>
+
+/**
+ * The ucs_rwlock_t type.
+ *
+ * Readers increment the counter by UCS_RWLOCK_READ (4)
+ * Writers set the UCS_RWLOCK_WRITE bit when lock is held
+ *     and set the UCS_RWLOCK_WAIT bit while waiting.
+ *
+ * 31                 2 1 0
+ * +-------------------+-+-+
+ * |  readers          | | |
+ * +-------------------+-+-+
+ *                      ^ ^
+ *                      | |
+ * WRITE: lock held ----/ |
+ * WAIT: writer pending --/
+ */
+
+#define UCS_RWLOCK_WAIT  0x1 /* Writer is waiting */
+#define UCS_RWLOCK_WRITE 0x2 /* Writer has the lock */
+#define UCS_RWLOCK_MASK  (UCS_RWLOCK_WAIT | UCS_RWLOCK_WRITE)
+                             /* Writer is waiting or has lock */
+#define UCS_RWLOCK_READ  0x4 /* Reader increment */
+
+
+/**
+ * Read-write lock.
+ */
+typedef struct {
+    volatile int l;
+} ucs_rwlock_t;
+
+
+static inline void ucs_rwlock_read_lock(ucs_rwlock_t *lock) {
+    int x;
+
+    while (1) {
+        while (lock->l & UCS_RWLOCK_MASK) {
+            sched_yield();
+        }
+
+        x = __sync_fetch_and_add(&lock->l, UCS_RWLOCK_READ);
```

Comment:
For my curiosity: why we are using deprecated atomic functions, while there are modern family of `__atomic_*` functions `__atomic_fetch_add`?
Those `__sync` functions always make a full barrier, and therefore come with significant overhead when only acquire or release semantics is sufficient

Reply:
From GCC documentation: https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html
`They [family of __sync function] should not be used for new code which should use the ‘__atomic’ builtins instead.`

Reply:
maybe consider using/extending `atomic.h`?